### ISSUE TITLE
Fix multiple issues in sujet_origine.py

### ISSUE THREAD

**Issue by aaryswastaken** [aaryswastaken/gandi-rush#19]

**Comment by aaryswastaken** [comment]
There has been an issue detected in the comprehension of the rules. 

The deletion of every connected values is only triggered when 3 are aligned horizontally or vertically

**Comment by aaryswastaken** [comment]
Fixing idea:

~~We could use the output of `detecte_coordonnees_combinaisons` to add a test if more than 3 are aligned next to eachother.~~
~~For that, an idea would be to use the non normalised output (using `iterations` rather than `out`) and to bruteforce the alignement test.~~
~~To counter this lack of optimization, when bruteforcing gravity, we could test only for non regiestered cases~~

Tho the last idea was really good, the best scenario would be:
 - On the user input, test only on the two affected cases (4 tests) 
 - On the gravity ticking, try every horizontal position then vertical position. Each time forgetting about border lines. This brute force method should be in worse case `O(2n)`, n representing the number of cell but we could use previous results to skip tests. For example:

Testing vertically: `1 1 2 1 3 3 3`
Removing borders: `- 1 2 1 3 3 -`
Testing for index 1: `1 1 2` -> wrong, we don't have to test for index 2 
Skipping to index 3: `2 1 3` -> wrong, can skip 4
Skipping to index 5: `3 3 3` -> Ok, exiting with True

This is reducing time complexity from `O(2n)` to `O(n)`

It would be interesting to find a `O(log(n))` solution for this tho

**Comment by aaryswastaken** [comment]
We could probably try to find an O(log n) by changing the grid storage and handling method and using graphs. 
Need to investigate on it and find suitable algorithms for gravity without the `graph -> grid -> transposed -> ticked -> untransposed -> updated_graph`

**Comment by aaryswastaken** [comment]
I have da solution

**Comment by aaryswastaken** [comment]
Ok so first of all imma write the "easy solution" to `sujet_origine.py` and do the advanced solution to the final version.
The goal is to create a class `CachedGraph`. `Grid.grid` will basically be an array of `CachedGraph`.

`CachedGraph` will be a set of cell (should consider the creation of the `Cell` class is necessary) and a reverse search array that maps each cell to the `Cell` that can refer itself to the `CachedGraph`.
This improves the `+` search (horizontal + vertical) because we only have to consider the search inside the graph. We can also ignore cells of the search for two reasons:
 - Is a border cell of the considered search (ie. x=0 for horizontal search) 
 - Is a border cell of the graph (the first case can be treated inside this one).

For example, lets consider a seemingly amorphous and random graph. 
`0=ok, 1=excluded from H, 2=excluded from V, 3=excluded from both`

We have: 
```
- - x x - x -             - - 3 3 - 3 -
- x x x x x -             - 3 2 0 2 3 - 
- - - x x - -      ->     - - - 1 3 - - 
- x x x - - -             - 3 2 3 - - - 
- - x - - - -             - - 3 - - - -
```

So if we want to search for horizontal matches during the physics tick (which used to be an expensive calculation), we only have to search for `0` and `2`.

The even better part is that the creation of such matrix is allowing us to filter because every graph containing at least one other value than 3 is going to be deleted.

We have to discuss if the computation of such set of graph is less expensive than the usual method.

Considering a square game board with `n` the size of such board:
 - The original computation of the graph is `O(n²)`
 - In the worse case scenario, the graph that is most at the bottom will be deleted. Detection and deletion will almost be `O(1)`
 - Because of the deletion, gravity ticking in its actual form will cause a re computation of the graph contained inside the deletion zone, and a merge with the border graphs. In the worse case, let `m`be the number of deleted cell and `M` the number of cell adjacent. Graph reconstruction will be `O(m)` (in the order of `O(n²)` because m is a cell number, tho `m < n²`) and graph merging should be at max `O(M)` with (`M << m` if `m` is big enough, otherwise those are minor operations so no problem either ways)
 - These steps should be repeated through gravity ticking

The total expensiveness is at most a complete graph reconstruction, costing `O(n²)` if everything is deleted.

If the grid entropy is large enough, a play with its associated ticking deletions will cost way less than the actual algorithm, meeting the `O(log n)` target.

We should also note that such graph implementation lets us see way faster if there is something tick-able, instead of testing every cell (`O(n²)`), we just have to search for `non-3` graphs.

Not simple but elegant in my opinion.

**Comment by aaryswastaken** [comment]
@Byllie can you review this pls

**Comment by Byllie** [comment]
Vraiment des mots compliqués pour des choses simples, basique. Je ne sais pas si inclure des graphs dans le projet est une bonne idée.

**Comment by aaryswastaken** [comment]
Hmmmm `O(n^2) -> O(log n)`

Fast as fuck boiiiiiiii

**Comment by aaryswastaken** [comment]
First addressed issue has been fixed through #23, now I'll focus on doing a little flamethrower job on the grid manager

**Comment by aaryswastaken** [comment]
Idea is too complex for that
